File: src/main/typescript/leetcode/_2_M.ts
class ListNode {
    val: number;
    next: ListNode;
    constructor(val: number) {
        this.val = val;
    }
}

function addTwoNumbers(l1: ListNode, l2: ListNode): ListNode {
    const tempHead = new ListNode(0);
    let curSum = tempHead;
    let cur1 = l1;
    let cur2 = l2;
    let additionCarry = 0;
    while (true) {
        let nulls = 0;
        if (cur1 != null) {
            additionCarry += cur1.val;
            cur1 = cur1.next;
        }
        else {
            nulls++;
        }
        if (cur2 != null) {
            additionCarry += cur2.val;
            cur2 = cur2.next;
        }
        else {
            nulls++;
        }

        if (nulls == 2 && additionCarry == 0) {
            break;
        }

        curSum.next = new ListNode(additionCarry % 10);
        curSum = curSum.next;
        additionCarry = Math.floor(additionCarry / 10);
    }
    return tempHead.next;
}

File: c++/src/leetcode/_2_M.cpp
class ListNode {
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(nullptr) {}
}

class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode tempNode(0);
        ListNode* curSum = &tempNode;
        ListNode* cur1 = l1;
        ListNode* cur2 = l2;
        int additionCarry = 0;
        while (true) {
            int breakFlag = 0;
            if (cur1 != nullptr) {
                additionCarry += cur1->val;
                cur1 = cur1->next;
            }
            else {
                breakFlag++;
            }
            if (cur2 != nullptr) {
                additionCarry += cur2->val;
                cur2 = cur2->next;
            }
            else {
                breakFlag++;
            }
            if (breakFlag == 2 && additionCarry == 0) {
                break;
            }
            curSum->next = new ListNode(additionCarry % 10);
            curSum = curSum->next;
            additionCarry /= 10;
        }
        return tempNode.next;
    }
};
